src/gainmap.c

```diff
@@ -49,4 +53,257 @@ avifBool avifGainMapMetadataFractionsToDouble(avifGainMapMetadataDouble * dst, c
+        avifDiagnosticsPrintf(diag,
+                              "Invalid gain map metadata, hdrCapacityMin should be <= hdrCapacityMax, got min %f and max %f",
+                              metadata.hdrCapacityMin,
+                              metadata.hdrCapacityMax);
+        return AVIF_RESULT_INVALID_ARGUMENT;
+    }
+    for (int i = 0; i < 3; ++i) {
+        if (metadata.gainMapGamma[i] <= 0) {
+            avifDiagnosticsPrintf(diag, "Invalid gain map metadata, gamma should be strictly positive");
+            return AVIF_RESULT_INVALID_ARGUMENT;
+        }
+    }
+
+    const uint32_t width = baseImage->width;
+    const uint32_t height = baseImage->height;
+    avifImage * rescaledGainMap = NULL;
+    avifRGBImage rgbGainMap;
+    // Basic zero-initialization for now, avifRGBImageSetDefaults() is called later on.
+    memset(&rgbGainMap, 0, sizeof(rgbGainMap));
+
+    avifResult res = AVIF_RESULT_OK;
+    toneMappedImage->width = width;
+    toneMappedImage->height = height;
+    AVIF_CHECKRES(avifRGBImageAllocatePixels(toneMappedImage));
+
+    // --- After this point, the function should exit with 'goto cleanup' to free allocated pixels.
+
+    const float weight = avifGetGainMapWeight(hdrCapacity, &metadata);
+
+    // Early exit if the gain map does not need to be applied and the pixel format is the same.
+    if (weight == 0.0f && outputTransferCharacteristics == transferCharacteristics && baseImage->format == toneMappedImage->format &&
+        baseImage->depth == toneMappedImage->depth && baseImage->isFloat == toneMappedImage->isFloat) {
+        assert(baseImage->rowBytes == toneMappedImage->rowBytes);
+        assert(baseImage->height == toneMappedImage->height);
+        // Copy the base image.
+        memcpy(toneMappedImage->pixels, baseImage->pixels, baseImage->rowBytes * baseImage->height);
+        goto cleanup;
+    }
+
+    avifRGBColorSpaceInfo baseRGBInfo;
+    avifRGBColorSpaceInfo toneMappedPixelRGBState;
+    if (!avifGetRGBColorSpaceInfo(baseImage, &baseRGBInfo) || !avifGetRGBColorSpaceInfo(toneMappedImage, &toneMappedPixelRGBState)) {
+        avifDiagnosticsPrintf(diag, "Unsupported RGB color space");
+        res = AVIF_RESULT_NOT_IMPLEMENTED;
+        goto cleanup;
+    }
+
+    const avifTransferFunction gammaToLinear = avifTransferCharacteristicsGetGammaToLinearFunction(transferCharacteristics);
+    const avifTransferFunction linearToGamma = avifTransferCharacteristicsGetLinearToGammaFunction(outputTransferCharacteristics);
+
+    // Early exit if the gain map does not need to be applied.
+    if (weight == 0.0f) {
+        // Just convert from one rgb format to another.
+        for (uint32_t j = 0; j < height; ++j) {
+            for (uint32_t i = 0; i < width; ++i) {
+                float basePixelRGBA[4];
+                avifGetRGBAPixel(baseImage, i, j, &baseRGBInfo, basePixelRGBA);
+                if (outputTransferCharacteristics != transferCharacteristics) {
```

Comment:
I agree this is a loop invariant but I think it's simpler/shorter this way? Not sure if there's a performance impact.

Reply:
Yes, let's go for simplicity. Branches in the inner loop may potentially hurt performance. There are if statements (branches) inside `avifGetRGBAPixel()` and `avifSetRGBAPixel()` too, so it doesn't seem necessary to avoid this if statement.